PaddlePaddle/Paddle
# make inference_api_test compile with dynamic linking library

### PR types

Others
### PR changes

Others
### Describe

Make inference_api_test compile with shared lib in windows

- make unittests in paddle/inference/tests/api link paddle_inference_shared and paddle_inference_c_shared rather than 
a list of static libraries( paddle_inference_api, paddle_inference_io, ir_pass_manager, analysis_predictor, benchmark), and that is what will happen in linux.
- make paddle_inference_shared in the begining of all dependencies. Because, the order of linking libraries, depending on the order of dependencies, may influence the .obj included by the executable program. For example, if put paddle_inference_shared in the end, the compiler will find the symbols needed in the static lib such as libzero_copy_tensor.lib and then include corresponding zero_copy_tensor.cc.obj. But paddle_inference.dll also contains symbols from zero_copy_tensor.cc.obj, and then conflict happens. If put paddle_inference_shared in the begining, the zero_copy_tensor.cc.obj will not be included and problem gone. In short, .dll must be linked, but .obj may not, so we should put the static lib in the back of shared lib.
![image](https://user-images.githubusercontent.com/51314274/175458712-e602a2f1-0969-43cf-a823-ef2e3ab69fc9.png)
- remove PADDLE_API in place.h, it not, symbols redefinition happen and cant be fixed. According to the CI result, the modification seems not have influence on linux.
![image](https://user-images.githubusercontent.com/51314274/175459346-fc54d7b7-3a3b-4300-bfad-8d1bbc92bbd5.png)
- make inference ut link paddle_inference_c_shared in windows and paddle_inference_c in linux. In windows, if still link paddle_inference_c in windows, symbols redefinition happen. While in linux, it change to paddle_inference_c_shared, error below happened.
![image](https://user-images.githubusercontent.com/51314274/175459149-59c6ea65-cc16-4d69-9978-c17df00eb751.png)
- compile paddle_inference_c_shared even when ON_INFER=OFF, since inference ut need to link paddle_inference_c_shared in windows.
- 4 *_test.cc get changed. These tests will fail before due to .exe cant get device context pool initialized in .dll, so .exe must initialize its own pool before calling SetFakeImageInput which will call executor->run.

### 影响：
#### 动态库导出符号变化：
Windows:
- paddle_inference.dll由develop的922个符号变为PR的3085个符号，不再包含GetOrRegisterGlobalDeviceTypeId和GetGlobalDeviceType。
- core_avx.pyd由develop的1242个符号变为PR的1235符号，减少的额符号为GetOrRegisterGlobalDeviceTypeId，GetGlobalDeviceType，Exp_EnableUseGpuFp16@AnalysisConfig，Exp_EnableUseGpuFp16@GpuPassStrategy，Exp_EnableUseGpuFp16@PassStrategy，gpu_fp16_enabled@AnalysisConfig，use_gpu_fp16@PassStrategy.

Linux:
- libpaddle_inference.so由develop的222562个符号变为PR的226852个符号，但是二者都包含了GetOrRegisterGlobalDeviceTypeId和GetGlobalDeviceType。
- core_avx.so由develop的272040个符号变为PR的270484个符号，但是二者都包含了GetOrRegisterGlobalDeviceTypeId和GetGlobalDeviceType。

导出符号文件压缩包下载  [链接](https://paddle-windows.bj.bcebos.com/share/symbols_changed.tar.gz)

#### 减小Windows平台编译产物体积：
在Windows-inference流水线上，Paddle/build/paddle/fluid/inference/tests/api目录下的预测单测体积由335MB减小为10MB，build目录总体积由37.3GB减小为18.6G。
before:
<img width="400" alt="image" src="https://user-images.githubusercontent.com/51314274/175899961-9aa24b42-5ce0-4b58-aa2c-15b309274314.png">

after:
<img width="400" alt="image" src="https://user-images.githubusercontent.com/51314274/175929804-e7d7feb0-c180-43f8-9418-06fbf2976805.png">

由于编译目标产物体积大大减小，能够从根本上解决并行编译时占用内存过多导致的编译卡、慢、link out of memory问题。经统计在windows-inference流水线，链接预测单测的总时间由1min减少到8s，若在32G物理内存机器上，效果将更加明显。
before:
并行编译时，会开多个link.exe进程同时执行链接操作，每个link.exe提交的内存修改都是4G，所需内存超过实际内存，开始使用虚拟内存，也就是固态硬盘，读写速度会下降，硬中断次数很多，导致编译时间长，甚至卡死，最终 link out of memory。
<img width="400" alt="image" src="https://user-images.githubusercontent.com/51314274/177297492-c8817447-1235-481e-b75a-708183a3a9f5.png">
<img width="400" alt="image" src="https://user-images.githubusercontent.com/51314274/177296544-47d3e452-9044-4b3b-ac4a-d8758b8d0992.png">

after：
在链接预测单测时，巅峰内存不超过20G。该部分编译不再成为瓶颈，Windows-CI对于机器的硬件配置要求降低，在32G物理内存的机器上，也能够流畅运行任务。
<img width="400" alt="image" src="https://user-images.githubusercontent.com/51314274/177296034-486d9005-8826-4dad-96cb-7ee1491f396a.png">
